src/typed_ast.hpp

```diff
@@ -166,90 +168,90 @@ struct Identifier : public TypedAST {
 
 // the value depends on the return value of callee
 struct CallExpression : public TypedAST {
-	Own<TypedAST> m_callee;
-	std::vector<Own<TypedAST>> m_args;
+	TypedAST* m_callee;
+	std::vector<TypedAST*> m_args;
 
 	CallExpression()
 	    : TypedAST {TypedASTTag::CallExpression} {}
 };
 
 struct IndexExpression : public TypedAST {
-	Own<TypedAST> m_callee;
-	Own<TypedAST> m_index;
+	TypedAST* m_callee;
+	TypedAST* m_index;
 
 	IndexExpression()
 	    : TypedAST {TypedASTTag::IndexExpression} {}
 };
 
 struct RecordAccessExpression : public TypedAST {
-	Own<TypedAST> m_record;
+	TypedAST* m_record;
 	Token const* m_member;
 
 	RecordAccessExpression()
 	    : TypedAST {TypedASTTag::RecordAccessExpression} {}
 };
 
 struct TernaryExpression : public TypedAST {
-	Own<TypedAST> m_condition;
-	Own<TypedAST> m_then_expr;
-	Own<TypedAST> m_else_expr;
+	TypedAST* m_condition;
+	TypedAST* m_then_expr;
+	TypedAST* m_else_expr;
 
 	TernaryExpression()
 	    : TypedAST {TypedASTTag::TernaryExpression} {}
 };
 
 struct Block : public TypedAST {
-	std::vector<Own<TypedAST>> m_body;
+	std::vector<TypedAST*> m_body;
 
 	Block()
 	    : TypedAST {TypedASTTag::Block} {}
 };
 
 struct ReturnStatement : public TypedAST {
-	Own<TypedAST> m_value;
+	TypedAST* m_value;
 
 	ReturnStatement()
 	    : TypedAST {TypedASTTag::ReturnStatement} {}
 };
 
 struct IfElseStatement : public TypedAST {
-	Own<TypedAST> m_condition;
-	Own<TypedAST> m_body;
-	Own<TypedAST> m_else_body; // can be nullptr
+	TypedAST* m_condition;
+	TypedAST* m_body;
+	TypedAST* m_else_body {nullptr}; // can be nullptr
 
 	IfElseStatement()
 	    : TypedAST {TypedASTTag::IfElseStatement} {}
 };
 
 struct ForStatement : public TypedAST {
-	Own<TypedAST> m_declaration;
-	Own<TypedAST> m_condition;
-	Own<TypedAST> m_action;
-	Own<TypedAST> m_body;
+	TypedAST* m_declaration;
+	TypedAST* m_condition;
+	TypedAST* m_action;
+	TypedAST* m_body;
 
 	ForStatement()
 	    : TypedAST {TypedASTTag::ForStatement} {}
 };
 
 struct WhileStatement : public TypedAST {
-	Own<TypedAST> m_condition;
-	Own<TypedAST> m_body;
+	TypedAST* m_condition;
+	TypedAST* m_body;
 
 	WhileStatement()
 	    : TypedAST {TypedASTTag::WhileStatement} {}
 };
 
 struct StructExpression : public TypedAST {
-	std::vector<Identifier> m_fields;
-	std::vector<Own<TypedAST>> m_types;
+	std::vector<Identifier*> m_fields;
```

Comment:
why are these pointers now?

Reply:
I figured since they are now allocated elsewhere I can't store them here directly, another approach would be to exclude some of the nodes in the node allocator and have them stored here wherever they are needed.